cli/src/pcluster/cli/entrypoint.py

```diff
@@ -1,155 +1,236 @@
-            epilog='For command specific flags, please run: "pcluster [command] --help"',
-        )
-        subparsers = self._parser.add_subparsers(dest="command")
-        subparsers.required = True
-        for command_class in CliCommand.__subclasses__() + CliCommandV3.__subclasses__():
-            if not isabstract(command_class):
-                command_class(subparsers)  # pylint: disable=abstract-class-instantiated
-
-        dcv = subparsers.add_parser(
-            "dcv",
-            help="The dcv command permits to use NICE DCV related features.",
-            epilog='For dcv subcommand specific flags, please run: "pcluster dcv [subcommand] --help"',
-        )
-        dcv_subparsers = dcv.add_subparsers(dest="command")
-        dcv_subparsers.required = True
-        for dcv_command_class in DcvCommand.__subclasses__():
-            dcv_command_class(dcv_subparsers)
-
-    def handle_command(self):
-        """Handle CLI command execution."""
+class APIOperationException(Exception):
+    def __init__(self, data):
+        self.data = data
+
+
+def bool_converter(in_str):
+    """Takes a boolean string and converts it into a boolean value."""
+    return in_str not in {'false', 'False', 'FALSE', False}
+
+
+def re_validator(rexp_str, param, in_str):
+    """Takes a string and validates the input format."""
+    rexp = re.compile(rexp_str)
+    if rexp.match(in_str) is None:
+        pprint({'message': f"Bad Request: '{in_str}' does not match '{rexp_str}' - '{param}'"})
+        sys.exit(1)
+    return in_str
+
+
+def read_file_b64(path):
+    """Takes file path, reads the file and converts to base64 encoded string"""
+    with open(path) as file:
+        file_data = file.read()
+    return base64.b64encode(file_data.encode('utf-8')).decode('utf-8')
+
+
+def convert_args(model, op_name, args_in):
+    """Takes a model, the name of the operation and the arguments
+    provided by argparse and converts the parameters into a format
+    that is suitable to be called in the controllers."""
+    body = {}
+    kwargs = {}
+    for param in model[op_name]['params']:
+        param_name = to_snake_case(param['name'])
+        value = args_in.pop(param_name)
+
+        if param['body']:
+            param_name = camelcase(param_name)
+            param_name = param_name[0].lower() + param_name[1:]
+            body[param_name] = value
+        else:
+            kwargs[param_name] = value
+
+    kwargs.update(args_in)
+
+    return body, kwargs
+
+
+def dispatch(model, args):
+    """Dispatches to a controller function when the arguments have an
+    operation specified."""
+    args_dict = args.__dict__
+    operation = args.operation
+    del args_dict['func']
+    del args_dict['operation']
+    body, kwargs = convert_args(model, operation, args_dict)
+
+    dispatch_func = partial(pcluster.cli.model.call, model[operation]['func'])
+    if len(body) > 0:
+        dispatch_func = partial(dispatch_func, body)
```

Comment:
I do think it's still helpful to have the body around for the middleware path so that we don't have to thread that information through, however it is still possible to make the function call by adding the body from the spec to the kwargs.

I updated the model parsing to pull this out and the `convert_args` function to leverage it. Middleware still accepts the body but it's just a reference to the same dict so it's possible to read / modify the body directly.